packages/sol-cov/src/trace_collection_subprovider.ts

```diff
@@ -0,0 +1,258 @@
+        if (this._isEnabled) {
+            switch (payload.method) {
+                case 'eth_sendTransaction':
+                    if (!this._config.shouldCollectTransactionTraces) {
+                        next();
+                    } else {
+                        const txData = payload.params[0];
+                        next(this._onTransactionSentAsync.bind(this, txData));
+                    }
+                    return;
+
+                case 'eth_call':
+                    if (!this._config.shouldCollectCallTraces) {
+                        next();
+                    } else {
+                        const callData = payload.params[0];
+                        next(this._onCallExecutedAsync.bind(this, callData));
+                    }
+                    return;
+
+                case 'eth_estimateGas':
+                    if (!this._config.shouldCollectGasEstimateTraces) {
+                        next();
+                    } else {
+                        const estimateGasData = payload.params[0];
+                        next(this._onGasEstimateAsync.bind(this, estimateGasData));
+                    }
+                    return;
+
+                default:
+                    next();
+                    return;
+            }
+        } else {
+            next();
+            return;
+        }
+    }
+    /**
+     * Set's the subprovider's engine to the ProviderEngine it is added to.
+     * This is only called within the ProviderEngine source code, do not call
+     * directly.
+     */
+    public setEngine(engine: Provider): void {
+        super.setEngine(engine);
+        this._web3Wrapper = new Web3Wrapper(engine);
+    }
+    private async _onTransactionSentAsync(
+        txData: MaybeFakeTxData,
+        err: Error | null,
+        txHash: string | undefined,
+        cb: Callback,
+    ): Promise<void> {
+        if (!txData.isFakeTransaction) {
+            // This transaction is a usual ttransaction. Not a call executed as one.
+            // And we don't want it to be executed within a snapshotting period
+            await this._lock.acquire();
+        }
+        if (_.isNull(err)) {
+            const toAddress = _.isUndefined(txData.to) || txData.to === '0x0' ? constants.NEW_CONTRACT : txData.to;
+            await this._recordTxTraceAsync(toAddress, txData.data, txHash as string);
+        } else {
+            const latestBlock = await this._web3Wrapper.getBlockWithTransactionDataAsync(BlockParamLiteral.Latest);
+            const transactions = latestBlock.transactions;
+            for (const transaction of transactions) {
+                const toAddress = _.isUndefined(txData.to) || txData.to === '0x0' ? constants.NEW_CONTRACT : txData.to;
+                await this._recordTxTraceAsync(toAddress, transaction.input, transaction.hash);
+            }
+        }
+        if (!txData.isFakeTransaction) {
```

Comment:
This conditional is duplicated. Why?

Reply:
First one is to acquire the lock. Second one is to release it.